pyproject.toml

```diff
@@ -77,7 +77,6 @@ dependencies = [
   "mkdocstrings~=0.22.0",
   "markdown-exec[ansi]~=1.6.0"
 ]
-detached = true
 type = "pip-compile"
```

Comment:
🧵 on environment inheritance

Reply:
Than you for all of the context about inheritance above @oprypin, I knew I was missing something there.

> Hatch environments have inheritance of all values. Any environment can choose to inherit from another environment. Multiple levels of inheritance are possible.

Does these mean that I would need to traverse up the tree of templates to determine which lock files to use as constraints files then @oprypin?

Reply:
> Does these mean that I would need to traverse up the tree of templates to determine which lock files to use as constraints files then?

Probably, or rather it would probably mean nested lock file dependencies.

Reply:
Of course I guess nobody is going to actually ever use these features this way, but hmm 🥲

Reply:
By the way, even in the current implementation, how does it work? You calculate that the `default` environment's lock file should be used as the constraint, but what if that lock file doesn't exist yet, what happens then? Should it force Hatch to calculate that environment first?

Reply:
Also I was saying that you shouldn't check `detached` and said about `skip_install` instead. Actually that is not what we care about! Per the above discussion what you really want to know is what value of `template` this env has. And it seems that Hatch doesn't even provide that answer, you maybe have to recompute the inheritance chain yourself somehow.

Reply:
Maybe you can't get the value of `template` at all!!
https://github.com/pypa/hatch/blob/d666ae7b3dacded86faa96cc0454bb1e69132dc7/src/hatch/project/config.py#L520
It's possible that only a metadata plugin would have access to that

Reply:
Okay, from what I'm seeing the environment `type` doesn't get inherited from a `template` though, is that right?

Reply:
In the current implementation we calculate that the default lockfile should be used as a constraint and if it doesn't exist we raise an error. If it does exist but it's dependencies are out of sync then it will also raise an error. 

Originally I was inclined to have it generate the parent's lockfile too, but I wasn't sure what would happen when I called `self.virtualenv.platform` from the `parent` class returned so I decided to play it safe and just raise an error.

https://github.com/juftin/hatch-pip-compile/blob/87489ccaa38bd024e8eb275a42feb179a644acf2/hatch_pip_compile/plugin.py#L219-L236

Reply:
🤔 okay here's some of my assumptions so far:

1. As long as `hatch-pip-compile` is enabled on the default environment, the environment's lockfile should always be used as a `constraints` file to another environment
    - Except when the `default` env has `skip_install` (AKA `detached`) set to `true` then it should not be used as a constraints
2. When an environment has `dependencies` unset it should use the constraints files from it's parent (whichever closest parent explicitly has `dependencies` set)
3. When an environment has `dependencies-extra` set it should use the lockfile of its parent

Reply:
> Okay, from what I'm seeing the environment `type` doesn't get inherited from a `template` though, is that right?

It's probably because you're setting `detached = true`. Otherwise it is inherited. It seems like setting `detached = true` destroys the `template` setting. So, setting both `detached` and `template` is nonsense, Hatch should change that behavior to perhaps warn that it  doesn't do anything.

Reply:
@juftin I think I'm forming a particular opinion on this now.

1. We can't feasibly follow the inheritance trail (as that config's value is not preserved)

2. I think I would not like error conditions to be possible per the default behavior of this plugin, and instead would prefer to not use the constraint functionality. re:
    > In the current implementation we calculate that the default lockfile should be used as a constraint and if it doesn't exist we raise an error.

With this in mind, why not make an explicit "constraint" setting - people would have to set something like `constraint = "default"`. With this:

* the setting is possible to control.
* people will know from the docs what they are opting into.
* you don't need to guess the desired lock file name.

Also funnily enough, this setting *would* be inherited,  I guess, so if people want to opt into locking every environment to "default", they could probably write this:

```toml
[envs.default]
constraint = "default"
```

Reply:
> As long as hatch-pip-compile is enabled on the default environment, the environment's lockfile should always be used as a constraints file to another environment 

I think that's not right! Any environment can be independent from the default environment just by inheriting from something else 

> When an environment has dependencies unset it should use the constraints files from it's parent 

Probably. With the caveat that I don't think there's any way to determine the parent?? The `template` config gets destroyed upon resolution

Reply:
Okay, this feedback is super useful. Trying all that recursive checking was exhausting, just setting the constraint env explicitly is a much better option. Here's where I've landed:

```toml
[tool.hatch.envs.default]
pip-compile-constraint = "default"
type = "pip-compile"
```

```toml
[tool.hatch.envs.docs]
dependencies = [
  "mkdocs~=1.5.2",
  "mkdocs-material~=9.2.3",
]
```

Boom 💥 that's it. `hatch-pip-compile` will always use the default environment's lockfile as a constraints file to environments that inherit from it (or explicitly set their own `hatch-pip-compile` environment).

Here's an example lockfile: https://github.com/juftin/hatch-pip-compile/blob/e7c4f5bfae822780d9e0a99dbc4bcd97442ed3a7/requirements/requirements-docs.txt